aiohttp/web_urldispatcher.py

```diff
@@ -979,8 +986,23 @@ def __init__(self) -> None:
         super().__init__()
         self._resources: List[AbstractResource] = []
         self._named_resources: Dict[str, AbstractResource] = {}
+        self._resource_index: dict[str, list[AbstractResource]] = {}
 
     async def resolve(self, request: Request) -> UrlMappingMatchInfo:
+        url_parts = request.rel_url.raw_parts
+        resource_index = self._resource_index
+        # Walk the url parts looking for candidates
+        for i in range(len(url_parts), 0, -1):
+            url_part = "/" + "/".join(url_parts[1:i])
+            if (resource_candidates := resource_index.get(url_part)) is not None:
+                for candidate in resource_candidates:
+                    if (
+                        match_dict := (await candidate.resolve(request))[0]
+                    ) is not None:
+                        return match_dict
```

Comment:
I think this changes the resolution order, which could cause problems for a few users, if they were depending on the resolution order previously.

I wonder if we should define the resolution order differently in v4, such as using the most specific match first (which would be similar to several other web standards).

Reply:
It walks it backwards to ensure the most exact url is resolved first, and if they have multiple registered at the same path, it uses a list so they should still be resolved in order.

I guess its possible that someone could register

`/everything` and `/everything/hidden` and never want `/everything/hidden` to work but that seems like it would be a bug?

Reply:
> It walks it backwards

Oh yes, I missed the -1.

We probably want to include some new tests to validate the ordering.

> /everything and /everything/hidden and never want /everything/hidden to work but that seems like it would be a bug?

You're probably right, I was thinking about variable routes, but I can't think of a situation where it'd cause a problem.

Reply:
I added a comment to better explain whats going on.  I'll also add some more tests for shadowing

Reply:
> You're probably right, I was thinking about variable routes, but I can't think of a situation where it'd cause a problem.

Probably doesn't matter in real usage, but it'd possible to have something like `/{x:.*/b}` vs `/x/{b}/` (where a request to /x/b/ would only be the same if the user had previously put the second route first).

Reply:
I pulled a few from https://docs.aiohttp.org/en/stable/web_quickstart.html#variable-resources

Reply:
Unfortunately the test fails https://github.com/aio-libs/aiohttp/pull/7829/files#r1393194544

I think this PR is DOA

Reply:
That test shows the variable path shadowing a fixed path. As you said previously, that'd probably be a user mistake, why would they add an endpoint that can't be accessed?

Reply:
The only concern is the edge case I showed above, which passes in your test, as you've ordered the /x/{b}/ route first, but would fail if you ordered it the other way on master. It's enough of an edge case that we can probably accept that risk (but, I'll probably push it back to the 3.10 release).

Reply:
The fixed paths are always going to win out over the dynamic paths. I guess thats OK as a breaking change since I don't think anyone will actually do that in production

Reply:
I adjusted the test to comment, and adjusted the opening text above to reflect the change in behavior.

Happy to wait for this to land in 4.0 as that seems fine for a behavior change.

Reply:
Also will need to adjust https://docs.aiohttp.org/en/stable/web_reference.html#resource

Reply:
adjusted that in https://github.com/aio-libs/aiohttp/pull/7829/commits/edf2f98c5eaa97547db5ff292940a1545cc54523

Will also add a note on fixed over dynamic

Reply:
> The fixed paths are always going to win out over the dynamic paths.

Another case I thought of last night, this doesn't hold with something like `/{x}/{y}/z` vs `/{x}/y/z`. In this case they'll be in the same list and end up back to the old behaviour.

I'm not sure if this is worth fixing, as it'll probably complicate the code further for a minimal edge case, but may affect how we document this...

Reply:
I'll adjust the docs later today

Reply:
I was too tired to finish this yesterday, but I've just pushed a change to the docs